### PR TITLE
Np 10629/workspace stripping

### PR DESCRIPTION
**Pull request by StigNorland**

### CODE REVIEW

File: src/main/java/no/sikt/sws/IndexHandler.java [hunk 1]
```diff
@@ -36,7 +36,9 @@ protected ProxyResponse<String> processProxyInput(
 
         var resourceIdentifier = request.getPathParameter(RESOURCE_IDENTIFIER);
 
-        if (resourceIdentifier.startsWith("_") || !resourceIdentifier.matches(ALLOWED_INPUT)) {
+        if (!"_alias".equals(resourceIdentifier)
+            &&  resourceIdentifier.startsWith("_")
+            || !resourceIdentifier.matches(ALLOWED_INPUT)) {
```
Comment (sondrev) [inline src/main/java/no/sikt/sws/IndexHandler.java:41]: This starts getting hard to read. Should be extracted to a method. We also shouldent use ! and && as its hard to follow.
Would be easier if it was like:
shouldThrowExcaption = reason1 || reason2 || reason3

File: src/main/java/no/sikt/sws/IndexHandler.java [hunk 2]
```diff
@@ -47,10 +49,15 @@ protected ProxyResponse<String> processProxyInput(
 
         try {
             var url = "/" + WorkspaceStripper.prefixUrl(resourceIdentifier, workspace);
+            var requestbody = body;
+
+            if ("_alias".equals(resourceIdentifier)) {
+                requestbody = WorkspaceStripper.prefixAliasBody(body, workspace);
```
Comment (sondrev) [inline src/main/java/no/sikt/sws/IndexHandler.java:55]: On line 52 you do requestbody = body, which is never used before its set to seomthing else.

File: src/main/java/no/sikt/sws/WorkspaceStripper.java [hunk 3]
```diff
@@ -15,6 +15,7 @@
 public class WorkspaceStripper {
 
     private static final Logger logger = LoggerFactory.getLogger(WorkspaceStripper.class);
+    public static final String INDEX = "_index";
```
Comment (sondrev) [inline src/main/java/no/sikt/sws/WorkspaceStripper.java:18]: We should find a more readable name for this. What is it used for? Its clearly not a index
Is it the name of the index-field in json? Maybe something like INDEX_FIELD_NAME would be better?

File: src/main/java/no/sikt/sws/WorkspaceStripper.java [hunk 4]
```diff
@@ -29,16 +30,20 @@ public static String remove(String body, String workspace) {
     }
 
     public static String prefixUrl(String index, String workspace) {
+        logger.info("prefixing " + workspace);
         if (index == null) {
-            return null;
+            return workspace + "-*";
+
+        } else if (index.startsWith("_")) {
+            return index;
```
Comment (sondrev) [inline src/main/java/no/sikt/sws/WorkspaceStripper.java:38]: As we dont allow _ actions after Fredriks changes today - is this check needed?

Comment (sondrev) [reply src/main/java/no/sikt/sws/WorkspaceStripper.java:38]: If you remove this then the whole method could be simplified:



var affix = (index == null) ? "*" : index
return workspace + "-" + affix

File: src/main/java/no/sikt/sws/WorkspaceStripper.java [hunk 5]
```diff
@@ -29,16 +30,20 @@ public static String remove(String body, String workspace) {
...
         }
-        logger.info("prefixing " + workspace);
         return workspace + "-" + index;
     }
 
     // replace {index} with {workspace}-{index} from responseBody
-    public static String prefixBody(String body, String workspace, String index) {
-        if (body == null || index == null) {
+    public static String prefixIndexesBody(String body, String workspace, String index) {
+        if (body == null || index == null || workspace == null) {
```
Comment (sondrev) [inline src/main/java/no/sikt/sws/WorkspaceStripper.java:46]: Do we need to log what was missing before returning null? Or would it be obvious if something fails?

File: src/main/java/no/sikt/sws/WorkspaceStripper.java [hunk 6]
```diff
@@ -49,19 +54,37 @@ public static String prefixBody(String body, String workspace, String index) {
     }
 
 
-    public static String prefixBody(List<JsonNode> gatewayBody, String workspaceprefix) {
+    public static String prefixIndexesBulkBody(List<JsonNode> gatewayBody, String workspacePrefix) {
+        if (gatewayBody == null || workspacePrefix == null) {
+            return null;
```
Comment (sondrev) [inline src/main/java/no/sikt/sws/WorkspaceStripper.java:59]: Do we need to log what was missing before returning null? Or would it be obvious if something fails?

File: src/test/java/no/sikt/sws/WorkspaceStripperTest.java [hunk 7]
```diff
@@ -91,61 +89,76 @@ Stream<DynamicTest> testPrefixBodyFactory() {
...
-    void assertBulkBodyPrefixing(TestCaseSws testCase) {
-        var gatewayBody = testCase.getRequestGateway().getBulkBody();
+    void assertBodyPrefixing(TestCaseSws testCase) {
+        var indexName = testCase.getIndexName();
         var expectedBody = testCase.getRequestOpensearch().getBody();
-        var resultBody = WorkspaceStripper.prefixBody(gatewayBody, WORKSPACEPREFIX);
 
-        assertEquals(expectedBody,resultBody);
+        if (indexName == null) {
+            logger.info("no index name, presuming bulk inserts");
```
Comment (sondrev) [inline src/test/java/no/sikt/sws/WorkspaceStripperTest.java:150]: instead of checking indexName == null here, you could iterate though request-bulk.json and for each run a assertBodyBulkPrefixing() and similar for the other test-types you create.

**Review by sondrev — CHANGES_REQUESTED** [review]

File: src/main/java/no/sikt/sws/IndexHandler.java [hunk 8]
```diff
@@ -35,33 +40,52 @@ protected ProxyResponse<String> processProxyInput(
...
+
         try {
-            var url = WorkspaceStripper.prefixUrl(resourceIdentifier, workspace);
+            if (searchCommand.equals(OpenSearchCommand.NOT_IMPLEMENTED)
+                || searchCommand.equals(OpenSearchCommand.INVALID)) {
 
+                validateResourceIdentifier(resourceIdentifier);
+            }
+
+            var url = "/" + WorkspaceStripper.prefixUrl(workspace, resourceIdentifier);
```
Comment (sondrev) [inline src/main/java/no/sikt/sws/IndexHandler.java:56]: You need to remove the "/" as this has changes since your merge

File: src/main/java/no/sikt/sws/IndexHandler.java [hunk 9]
```diff
@@ -35,33 +40,52 @@ protected ProxyResponse<String> processProxyInput(
...
             logger.info("raw response-body:" + response.getBody());
 
-            var responseBody = WorkspaceStripper.remove(response.getBody(), workspace);
+            var responseBody = WorkspaceStripper.removePrefix(workspace, response.getBody());
             logger.info("stripped response-body:" + responseBody);
 
             return new ProxyResponse<>(response.getStatus(), responseBody);
+        } catch (BadRequestException be) {
+            logger.error(be.getLocalizedMessage());
+            throw be;
```
Comment (sondrev) [inline src/main/java/no/sikt/sws/IndexHandler.java:76]: Since we are at the IndexHandler we should throw it as a managed exception so we know what is returned to the caller. Perhaps use throw new SearchException(e.getMessage(), e); or something similar

File: src/test/java/no/sikt/sws/IndexHandlerTest.java [hunk 10]
```diff
@@ -55,13 +53,13 @@ void shouldGiveResponse() throws IOException {
 
         final OpenSearchResponse mockResponse = new OpenSearchResponse(200, "{}");
 
-        when(openSearchClient.sendRequest(PUT, TEST_WORKSPACE_PREFIX + TEST_INDEX_1 + "/_mapping", null))
+        when(openSearchClient.sendRequest(GET, "/" + TEST_WORKSPACE_PREFIX + TEST_INDEX_1 + "/_mapping", null))
```
Comment (sondrev) [inline src/test/java/no/sikt/sws/IndexHandlerTest.java:56]: no "/"

File: src/main/java/no/sikt/sws/IndexHandler.java (hunk 9 above)
Comment (StigNorland) [reply src/main/java/no/sikt/sws/IndexHandler.java:76]: ok

File: src/main/java/no/sikt/sws/WorkspaceStripper.java (hunk 5 above)
Comment (StigNorland) [reply src/main/java/no/sikt/sws/WorkspaceStripper.java]: -> Exception with message

File: src/main/java/no/sikt/sws/WorkspaceStripper.java (hunk 6 above)
Comment (StigNorland) [reply src/main/java/no/sikt/sws/WorkspaceStripper.java]: -> Exception with message

**Review by sondrev — APPROVED** [review]
